cardano-api/src/Cardano/Api/ProtocolParameters.hs

```diff
@@ -213,14 +204,347 @@ data ProtocolParameters =
+
+       -- | Price of execution units for non-native script languages.
+       protocolParamPrices :: Maybe ExecutionUnitPrices,
+
+       -- | Max total script execution resources units allowed per tx
+       protocolParamMaxTxExUnits :: Maybe MaxTxExecutionUnits,
+
+       -- | Max total script execution resources units allowed per block
+       protocolParamMaxBlockExUnits :: Maybe MaxBlockExecutionUnits,
+
+       -- | Max size of a Value in a tx ouput.
+       protocolParamMaxValSize :: Maybe Natural
     }
   deriving (Eq, Generic, Show)
 
-instance FromJSON ProtocolParameters where
-  parseJSON = withObject "ProtocolParameters" $ \o -> do
-                v <- o .: "protocolVersion"
-                ProtocolParameters
+data ProtocolParametersError
+  = ProtocolParamsErrNoParamsInByron
+  | ProtocolParamsErrMustDefineMinUTxo AnyCardanoEra
+  -- Alonzo required protocol params
+  | ProtocolParamsErrMustDefineCostPerByte AnyCardanoEra
+  | ProtocolParamsErrMustDefineCostModel AnyCardanoEra
+  | ProtocolParamsErrMustDefineScriptExecPrices AnyCardanoEra
+  | ProtocolParamsErrMustDefineMaxTxExec AnyCardanoEra
+  | ProtocolParamsErrMustDefineMaxBlockExec AnyCardanoEra
+  | ProtocolParamsErrMustDefineMaxValueSize AnyCardanoEra
+  -- Error Checking
+  | ProtocolParametersErrorWrongVersion
+      (Natural,Natural)
+      (Natural,Natural)
+  deriving Show
+
+renderEra :: AnyCardanoEra -> Text
+renderEra (AnyCardanoEra ByronEra)   = "Byron"
+renderEra (AnyCardanoEra ShelleyEra) = "Shelley"
+renderEra (AnyCardanoEra AllegraEra) = "Allegra"
+renderEra (AnyCardanoEra MaryEra)    = "Mary"
+
+renderProtocolParamsErr :: ProtocolParametersError -> Text
+renderProtocolParamsErr ProtocolParamsErrNoParamsInByron =
+  "Protocol parameters are not supported in the Byron era."
+renderProtocolParamsErr (ProtocolParamsErrMustDefineMinUTxo era) =
+  "protocolParamMinUTxOValue must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineCostPerByte era) =
+  "protocolParamUTxOCostPerByte must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineCostModel era) =
+  "protocolParamCostModels must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineScriptExecPrices era) =
+  "protocolParamPrices must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineMaxTxExec era) =
+  "protocolParamMaxTxExUnits must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineMaxBlockExec era) =
+  "protocolParamMaxBlockExUnits must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParamsErrMustDefineMaxValueSize era) =
+  "protocolParamMaxValSize must be defined in " <> renderEra era
+renderProtocolParamsErr (ProtocolParametersErrorWrongVersion expected got) =
+     "Incorrect protocol version. Got: " <> Text.pack (show got)
+  <> " Expected: " <> Text.pack (show expected)
+
+createProtocolParameters
```

Comment:
Why does this function exist? What does it add over using the `ProtocolParameters` constructor and all the nice field names?